File: pipeline/src/latentsky/manifest.py
"""Emit manifest.json conforming EXACTLY to schema/manifest.schema.json — §10.

The manifest is THE contract between pipeline/ and web/: the web app renders
only what this file declares, and the pipeline emits it LAST, after every
referenced asset exists on disk and every identity gate has passed. Order of
operations here is therefore load-bearing:

    1. verify_identity()       — coarse/fine ramp identities equal per variable
    2. verify_global_ranges()  — every layer on the GLOBAL vmin/vmax
    3. frame times strictly increasing, frame arrays the right length
    4. every referenced frame file and LUT exists
    5. jsonschema Draft 2020-12 validation against the committed schema
    6. only then write manifest.json

Any failure raises; nothing is written on a failed gate.
"""

from __future__ import annotations

import json
import pathlib

import jsonschema

from .encode import LayerRecord, verify_global_ranges, verify_identity
from .ramps import RampSpec

REPO_ROOT = pathlib.Path(__file__).resolve().parents[3]
SCHEMA_PATH = REPO_ROOT / "schema" / "manifest.schema.json"


class ManifestError(RuntimeError):
    """The manifest could not be emitted. The build must stop."""


def load_schema(schema_path: pathlib.Path = SCHEMA_PATH) -> dict:
    with open(schema_path, "r", encoding="utf-8") as fh:
        return json.load(fh)


_BASEMAP_KEYS = {"global", "globalRect", "hero", "heroRect"}

# Config key (snake_case, as in configs/event_*.yaml) -> schema run field (camelCase).
_HINT_KEYS = {
    "storm_name": "stormName",
    "hero_frame": "heroFrame",
    "place_label": "placeLabel",
    "report": "reportUrl",
    "default_variable": "defaultVariable",
}


def run_hints(cfg: dict) -> dict:
    """Extract the OPTIONAL first-class run hints from a config dict.

    Config-driven by design: event configs (configs/event_*.yaml) and the dev
    encode both route their {storm_name, hero_frame, place_label} through here,
    so the UI never regex-parses a storm name out of generatedNote or guesses
    the hero frame. Missing keys are simply absent from the result — the web
    app falls back to its heuristics. Wrong types raise, loudly: a silently
    dropped hint would present as the heuristic, which is exactly the failure
    class this field exists to remove. hero_frame bounds against the frame
    count are gated later, in build_manifest, where the frames are known.
    """
    hints: dict = {}
    for key, field in _HINT_KEYS.items():
        if key not in cfg or cfg[key] is None:
            continue
        value = cfg[key]
        if key == "hero_frame":
            if isinstance(value, bool) or not isinstance(value, int) or value < 0:
                raise ManifestError(
                    f"config {key}: expected a non-negative integer frame index, got {value!r}"
                )
        else:
            if not isinstance(value, str) or not value.strip():
                raise ManifestError(f"config {key}: expected a non-empty string, got {value!r}")
        hints[field] = value
    return hints


def build_manifest(
    run: dict,
    frame_times: list[str],
    layers: list[LayerRecord],
    specs: dict[str, RampSpec],
    basemap: dict | None = None,
) -> dict:
    """Assemble the manifest dict and run every pre-schema gate. Raises on any failure.

    `basemap` is the optional schema "basemap" object ({global, globalRect, hero,
    heroRect}); its imagery must be public-domain or CC-BY (schema note) and its
    referenced files are existence-checked by write_manifest like every layer asset.
    """
    if not layers:
        raise ManifestError("no layers — refusing to emit an empty manifest")

    if basemap is not None:
        unknown = set(basemap) - _BASEMAP_KEYS
        if unknown:
            raise ManifestError(f"basemap carries unknown keys {sorted(unknown)}")
        if not ("global" in basemap or "hero" in basemap):
            raise ManifestError("basemap object declares no imagery at all")
        for key in ("globalRect", "heroRect"):
            if key in basemap:
                rect = basemap[key]
                if len(rect) != 4 or not all(isinstance(v, (int, float)) for v in rect):
                    raise ManifestError(f"basemap.{key} must be [west, south, east, north], got {rect!r}")

    # Gates 1 and 2 — §7.2(b). These raise RampIdentityError / GlobalRangeError.
    verify_identity(layers)
    verify_global_ranges(layers, specs)

    # Gate 3 — frame bookkeeping.
    if any(a >= b for a, b in zip(frame_times, frame_times[1:])):
        raise ManifestError(f"frame times must be strictly increasing, got {frame_times}")
    hero_frame = run.get("heroFrame")
    if hero_frame is not None:
        if isinstance(hero_frame, bool) or not isinstance(hero_frame, int):
            raise ManifestError(f"run.heroFrame must be an integer frame index, got {hero_frame!r}")
        if not 0 <= hero_frame < len(frame_times):
            raise ManifestError(
                f"run.heroFrame {hero_frame} is out of range for {len(frame_times)} frames — "
                "the schema cannot cross-reference frames, so this gate lives here"
            )
    for layer in layers:
        if len(layer.frames) != len(frame_times):
            raise ManifestError(
                f"layer {layer.layer_id}: {len(layer.frames)} frames for "
                f"{len(frame_times)} frame times"
            )

    # A declared default variable must exist as a renderable layer, or the UI
    # would silently fall back and the config's intent would vanish without a
    # word. Gated here because a JSON schema cannot cross-reference the layers.
    wanted = run.get("defaultVariable")
    if wanted is not None:
        available = {layer.variable for layer in layers if layer.kind in ("global", "hero-fine")}
        if wanted not in available:
            raise ManifestError(
                f"run.defaultVariable {wanted!r} has no global or hero-fine layer in this run "
                f"(emitted: {sorted(available)})"
            )

    by_id = {layer.layer_id: layer for layer in layers}
    for layer in layers:
        if layer.pair_with is None:
            continue
        if layer.pair_with not in by_id:
            raise ManifestError(
                f"layer {layer.layer_id}: pairWith {layer.pair_with!r} is not a layer id"
            )
        # The viewer refuses a pair whose variables differ (one legend cannot
        # describe two ramps), and it does so at load time in the browser. Fail
        # here instead, where the build can see it.
        other = by_id[layer.pair_with]
        if other.variable != layer.variable:
            raise ManifestError(
                f"layer {layer.layer_id} ({layer.variable}) pairs with {other.layer_id} "
                f"({other.variable}) — a reveal must compare the same variable"
            )

    manifest: dict = {
        "schemaVersion": 1,
        "run": run,
        "frames": list(frame_times),
    }
    if basemap is not None:
        manifest["basemap"] = {
            k: (list(map(float, v)) if k.endswith("Rect") else str(v))
            for k, v in basemap.items()
        }
    manifest["layers"] = {}
    for layer in layers:
        entry: dict = {
            "kind": layer.kind,
            "variable": layer.variable,
            "label": layer.label,
            "units": layer.units,
            "rect": [float(x) for x in layer.rect],
            "size": [int(x) for x in layer.size],
            "lut": layer.lut,
            "vmin": float(layer.vmin),
            "vmax": float(layer.vmax),
            "identity": layer.identity,
            "frames": list(layer.frames),
        }
        if layer.pair_with is not None:
            entry["pairWith"] = layer.pair_with
        if layer.native_km is not None:
            entry["nativeKm"] = layer.native_km
        manifest["layers"][layer.layer_id] = entry
    return manifest


def write_manifest(
    manifest: dict,
    out_dir: pathlib.Path,
    schema_path: pathlib.Path = SCHEMA_PATH,
) -> pathlib.Path:
    """Gate 4 (assets exist) + gate 5 (schema) and only then write manifest.json."""
    out_dir = pathlib.Path(out_dir)

    missing: list[str] = []
    for key in ("global", "hero"):
        rel = manifest.get("basemap", {}).get(key)
        if rel is not None and not (out_dir / rel).is_file():
            missing.append(f"basemap.{key}: {rel}")
    for layer_id, layer in manifest["layers"].items():
        for rel in [layer["lut"], *layer["frames"]]:
            if not (out_dir / rel).is_file():
                missing.append(f"{layer_id}: {rel}")
    if missing:
        raise ManifestError(
            "manifest references assets that do not exist — the manifest must be emitted "
            "last:\n  " + "\n  ".join(missing)
        )

    schema = load_schema(schema_path)
    validator = jsonschema.Draft202012Validator(schema)
    errors = sorted(validator.iter_errors(manifest), key=lambda e: list(e.absolute_path))
    if errors:
        detail = "\n  ".join(
            f"at {'/'.join(str(p) for p in err.absolute_path) or '<root>'}: {err.message}"
            for err in errors
        )
        raise ManifestError(f"manifest fails schema validation:\n  {detail}")

    path = out_dir / "manifest.json"
    path.write_text(json.dumps(manifest, indent=2) + "\n", encoding="utf-8")
    return path

File: web/src/data/manifest.ts
/**
 * Typed loader for the data manifest — THE contract between pipeline/ and web/
 * (schema/manifest.schema.json). Lightweight structural validation, not full
 * jsonschema: every check throws with the JSON path of the violation, because a
 * malformed manifest must fail loudly, never render half a globe.
 */

export type Variable = "wind10m" | "t2m" | "mrr" | "tcwv" | "msl";
export type LayerKind = "global" | "hero-fine" | "hero-coarse";

const VARIABLES: readonly Variable[] = ["wind10m", "t2m", "mrr", "tcwv", "msl"];
const KINDS: readonly LayerKind[] = ["global", "hero-fine", "hero-coarse"];

export interface RunInfo {
  id: string;
  kind: "forecast" | "dev-sample";
  init?: string;
  model: { prognostic: string; downscaling: string };
  generatedNote: string;
  /** First-class storm name, e.g. "Typhoon Gaemi". Absent when no storm claim
   * should be made — use stormNameFor() for the heuristic fallback. */
  stormName?: string;
  /** Index into frameIso of the most developed frame — where "enter the storm"
   * should arrive. Use heroFrameFor() for the last-frame fallback. */
  heroFrame?: number;
  /** Human label for the hero domain, e.g. "Taiwan · CWA model domain". */
  placeLabel?: string;
}

export interface LayerDef {
  id: string;
  kind: LayerKind;
  variable: Variable;
  label: string;
  units: string;
  /** [west, south, east, north] degrees */
  rect: [number, number, number, number];
  /** [width, height] pixels */
  size: [number, number];
  /** Absolute URL of the shared 256x1 RGBA LUT PNG */
  lutUrl: string;
  vmin: number;
  vmax: number;
  identity: string;
  /** Absolute URL per frame, same length and order as Manifest.frameIso */
  frameUrls: string[];
  pairWith?: string;
}

/**
 * Optional offline-baked basemap imagery (schema `basemap`). URLs are resolved
 * against the manifest URL, exactly like layer frames. Arrives pre-styled —
 * §7.2(c): the app never touches brightness/contrast/hue/saturation/gamma.
 */
export interface BasemapDef {
  /** Absolute URL of the full-globe equirectangular basemap, if declared. */
  globalUrl?: string;
  /** [west, south, east, north] degrees; defaults to the full globe. */
  globalRect: [number, number, number, number];
  /** Absolute URL of the crisp regional basemap over the hero rectangle, if declared. */
  heroUrl?: string;
  heroRect?: [number, number, number, number];
}

export interface Manifest {
  schemaVersion: 1;
  run: RunInfo;
  /** Valid times as ISO strings, strictly increasing */
  frameIso: string[];
  basemap?: BasemapDef;
  layers: Map<string, LayerDef>;
}

export function manifestUrlFromLocation(loc: Location = window.location): string {
  return new URLSearchParams(loc.search).get("manifest") ?? "/data/web/manifest.json";
}

function fail(path: string, message: string): never {
  throw new Error(`manifest invalid at ${path}: ${message}`);
}

function requireString(v: unknown, path: string): string {
  if (typeof v !== "string" || v.length === 0) fail(path, `expected non-empty string, got ${JSON.stringify(v)}`);
  return v;
}

function requireNumber(v: unknown, path: string): number {
  if (typeof v !== "number" || !Number.isFinite(v)) fail(path, `expected finite number, got ${JSON.stringify(v)}`);
  return v;
}

function requireRecord(v: unknown, path: string): Record<string, unknown> {
  if (typeof v !== "object" || v === null || Array.isArray(v)) fail(path, "expected object");
  return v as Record<string, unknown>;
}

function requireIndex(v: unknown, path: string): number {
  const n = requireNumber(v, path);
  if (!Number.isInteger(n) || n < 0) fail(path, `expected non-negative integer index, got ${n}`);
  return n;
}

function requireIso(v: unknown, path: string): string {
  const s = requireString(v, path);
  if (Number.isNaN(Date.parse(s))) fail(path, `not a parseable date-time: ${s}`);
  return s;
}

function requireRect(v: unknown, path: string): [number, number, number, number] {
  if (!Array.isArray(v) || v.length !== 4) fail(path, "expected [west, south, east, north]");
  const rect = v.map((n, i) => requireNumber(n, `${path}[${i}]`)) as [number, number, number, number];
  if (!(rect[0] < rect[2] && rect[1] < rect[3])) fail(path, `degenerate rectangle ${JSON.stringify(rect)}`);
  return rect;
}

export async function loadManifest(url: string): Promise<Manifest> {
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`manifest fetch failed: ${response.status} ${response.statusText} for ${url}`);
  }
  const raw: unknown = await response.json();
  const baseUrl = new URL(url, window.location.href);
  return parseManifest(raw, baseUrl);
}

export function parseManifest(raw: unknown, baseUrl: URL): Manifest {
  const root = requireRecord(raw, "$");
  if (root.schemaVersion !== 1) fail("$.schemaVersion", `expected 1, got ${JSON.stringify(root.schemaVersion)}`);

  const runRaw = requireRecord(root.run, "$.run");
  const kind = requireString(runRaw.kind, "$.run.kind");
  if (kind !== "forecast" && kind !== "dev-sample") fail("$.run.kind", `unknown kind ${kind}`);
  const modelRaw = requireRecord(runRaw.model, "$.run.model");
  const run: RunInfo = {
    id: requireString(runRaw.id, "$.run.id"),
    kind,
    init: runRaw.init === undefined ? undefined : requireIso(runRaw.init, "$.run.init"),
    model: {
      prognostic: requireString(modelRaw.prognostic, "$.run.model.prognostic"),
      downscaling: requireString(modelRaw.downscaling, "$.run.model.downscaling"),
    },
    generatedNote: requireString(runRaw.generatedNote, "$.run.generatedNote"),
    stormName: runRaw.stormName === undefined ? undefined : requireString(runRaw.stormName, "$.run.stormName"),
    heroFrame: runRaw.heroFrame === undefined ? undefined : requireIndex(runRaw.heroFrame, "$.run.heroFrame"),
    placeLabel: runRaw.placeLabel === undefined ? undefined : requireString(runRaw.placeLabel, "$.run.placeLabel"),
  };

  if (!Array.isArray(root.frames) || root.frames.length < 1) fail("$.frames", "expected non-empty array");
  const frameIso = root.frames.map((f, i) => requireIso(f, `$.frames[${i}]`));
  for (let i = 1; i < frameIso.length; i++) {
    if (Date.parse(frameIso[i]) <= Date.parse(frameIso[i - 1])) {
      fail(`$.frames[${i}]`, `not strictly increasing: ${frameIso[i - 1]} -> ${frameIso[i]}`);
    }
  }
  // heroFrame indexes into frames — checkable only once both are parsed.
  if (run.heroFrame !== undefined && run.heroFrame >= frameIso.length) {
    fail("$.run.heroFrame", `index ${run.heroFrame} out of range for ${frameIso.length} frames`);
  }

  let basemap: BasemapDef | undefined;
  if (root.basemap !== undefined) {
    const b = requireRecord(root.basemap, "$.basemap");
    basemap = {
      globalUrl:
        b.global === undefined
          ? undefined
          : new URL(requireString(b.global, "$.basemap.global"), baseUrl).toString(),
      globalRect:
        b.globalRect === undefined ? [-180, -90, 180, 90] : requireRect(b.globalRect, "$.basemap.globalRect"),
      heroUrl:
        b.hero === undefined
          ? undefined
          : new URL(requireString(b.hero, "$.basemap.hero"), baseUrl).toString(),
      heroRect: b.heroRect === undefined ? undefined : requireRect(b.heroRect, "$.basemap.heroRect"),
    };
  }

  const layersRaw = requireRecord(root.layers, "$.layers");
  const ids = Object.keys(layersRaw);
  if (ids.length < 1) fail("$.layers", "expected at least one layer");

  const layers = new Map<string, LayerDef>();
  for (const id of ids) {
    const p = `$.layers[${JSON.stringify(id)}]`;
    const l = requireRecord(layersRaw[id], p);

    const layerKind = requireString(l.kind, `${p}.kind`) as LayerKind;
    if (!KINDS.includes(layerKind)) fail(`${p}.kind`, `unknown kind ${layerKind}`);
    const variable = requireString(l.variable, `${p}.variable`) as Variable;
    if (!VARIABLES.includes(variable)) fail(`${p}.variable`, `unknown variable ${variable}`);

    const rect = requireRect(l.rect, `${p}.rect`);

    if (!Array.isArray(l.size) || l.size.length !== 2) fail(`${p}.size`, "expected [width, height]");
    const size = l.size.map((v, i) => {
      const n = requireNumber(v, `${p}.size[${i}]`);
      if (!Number.isInteger(n) || n < 1) fail(`${p}.size[${i}]`, `expected positive integer, got ${n}`);
      return n;
    }) as [number, number];

    const vmin = requireNumber(l.vmin, `${p}.vmin`);
    const vmax = requireNumber(l.vmax, `${p}.vmax`);
    if (!(vmin < vmax)) fail(`${p}.vmin`, `vmin ${vmin} must be < vmax ${vmax}`);

    if (!Array.isArray(l.frames) || l.frames.length !== frameIso.length) {
      fail(`${p}.frames`, `expected ${frameIso.length} frame paths (one per $.frames entry), got ${Array.isArray(l.frames) ? l.frames.length : typeof l.frames}`);
    }
    const frameUrls = l.frames.map((f, i) => new URL(requireString(f, `${p}.frames[${i}]`), baseUrl).toString());

    layers.set(id, {
      id,
      kind: layerKind,
      variable,
      label: requireString(l.label, `${p}.label`),
      units: requireString(l.units, `${p}.units`),
      rect,
      size,
      lutUrl: new URL(requireString(l.lut, `${p}.lut`), baseUrl).toString(),
      vmin,
      vmax,
      identity: requireString(l.identity, `${p}.identity`),
      frameUrls,
      pairWith: l.pairWith === undefined ? undefined : requireString(l.pairWith, `${p}.pairWith`),
    });
  }

  // Cross-references: pairWith must resolve, and a pair must share its variable.
  for (const [id, layer] of layers) {
    if (layer.pairWith !== undefined) {
      const other = layers.get(layer.pairWith);
      if (!other) fail(`$.layers[${JSON.stringify(id)}].pairWith`, `no such layer ${layer.pairWith}`);
      if (other.variable !== layer.variable) {
        fail(`$.layers[${JSON.stringify(id)}].pairWith`, `pairs ${layer.variable} with ${other.variable} — a reveal must compare the same variable`);
      }
    }
  }

  return { schemaVersion: 1, run, frameIso, basemap, layers };
}

/**
 * The storm's display name. Prefers the first-class run.stormName; falls back
 * to the legacy heuristic (parsing "Typhoon X" out of generatedNote), then to
 * the honest non-claim "the hero region". UI adoption point for the invitation.
 */
export function stormNameFor(manifest: Manifest): string {
  return (
    manifest.run.stormName ??
    /Typhoon\s+[A-Z][a-z]+/.exec(manifest.run.generatedNote)?.[0] ??
    "the hero region"
  );
}

/**
 * The frame index "enter the storm" should arrive on. Prefers the first-class
 * run.heroFrame (already range-checked at parse); falls back to the legacy
 * heuristic — the last frame, on the grounds that frames are chronological and
 * quiet early frames must never sit under the invitation copy.
 */
export function heroFrameFor(manifest: Manifest): number {
  return manifest.run.heroFrame ?? manifest.frameIso.length - 1;
}

/**
 * Human label for the hero domain, e.g. "Taiwan · CWA model domain", or
 * undefined when the manifest makes no claim — there is no heuristic to fall
 * back to, and inventing a place would be exactly the failure the first-class
 * field exists to prevent.
 */
export function placeLabelFor(manifest: Manifest): string | undefined {
  return manifest.run.placeLabel;
}

/** The hero-fine layer for a variable, or undefined. */
export function heroFineLayerFor(manifest: Manifest, variable: Variable): LayerDef | undefined {
  for (const layer of manifest.layers.values()) {
    if (layer.variable === variable && layer.kind === "hero-fine") return layer;
  }
  return undefined;
}

/** The kind-"global" layer for a variable, or undefined. */
export function globalLayerFor(manifest: Manifest, variable: Variable): LayerDef | undefined {
  for (const layer of manifest.layers.values()) {
    if (layer.variable === variable && layer.kind === "global") return layer;
  }
  return undefined;
}

/** The hero-fine (falling back to global) layer for a variable, or undefined. */
export function primaryLayerFor(manifest: Manifest, variable: Variable): LayerDef | undefined {
  return heroFineLayerFor(manifest, variable) ?? globalLayerFor(manifest, variable);
}

/** Variables this manifest can actually render, in schema enum order. */
export function availableVariables(manifest: Manifest): Variable[] {
  return VARIABLES.filter((v) => primaryLayerFor(manifest, v) !== undefined);
}
